File: ta4j-core/src/test/java/org/ta4j/core/rules/InSlopeRuleTest.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.rules;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.ta4j.core.BarSeries;
import org.ta4j.core.BaseBarSeriesBuilder;
import org.ta4j.core.Indicator;
import org.ta4j.core.Rule;
import org.ta4j.core.TraceTestLogger;
import org.ta4j.core.indicators.helpers.FixedNumIndicator;
import org.ta4j.core.num.Num;

public class InSlopeRuleTest {

    private InSlopeRule rulePositiveSlope;
    private InSlopeRule ruleNegativeSlope;
    private BarSeries series;
    private TraceTestLogger traceTestLogger;

    @Before
    public void setUp() {
        series = new BaseBarSeriesBuilder().build();
        Indicator<Num> indicator = new FixedNumIndicator(series, 50, 70, 80, 90, 99, 60, 30, 20, 10, 0);
        rulePositiveSlope = new InSlopeRule(indicator, series.numFactory().numOf(20), series.numFactory().numOf(30));
        ruleNegativeSlope = new InSlopeRule(indicator, series.numFactory().numOf(-40), series.numFactory().numOf(-20));
        traceTestLogger = new TraceTestLogger();
        traceTestLogger.open();
    }

    @After
    public void tearDown() {
        traceTestLogger.close();
    }

    @Test
    public void isSatisfied() {
        assertFalse(rulePositiveSlope.isSatisfied(0));
        assertTrue(rulePositiveSlope.isSatisfied(1));
        assertFalse(rulePositiveSlope.isSatisfied(2));
        assertFalse(rulePositiveSlope.isSatisfied(9));

        assertFalse(ruleNegativeSlope.isSatisfied(0));
        assertFalse(ruleNegativeSlope.isSatisfied(1));
        assertTrue(ruleNegativeSlope.isSatisfied(5));
        assertFalse(ruleNegativeSlope.isSatisfied(9));
    }

    @Test
    public void traceIncludesSlopeValues() {
        assertTrue(rulePositiveSlope.isSatisfiedWithTraceMode(1, Rule.TraceMode.VERBOSE));

        String logContent = traceTestLogger.getLogOutput();
        assertTrue("Trace should include the current value", logContent.contains("currentValue=70"));
        assertTrue("Trace should include the previous value", logContent.contains("previousValue=50"));
        assertTrue("Trace should include the computed slope", logContent.contains("slope=20"));
        assertTrue("Trace should include the minimum slope", logContent.contains("minSlope=20"));
        assertTrue("Trace should include the maximum slope", logContent.contains("maxSlope=30"));
        assertTrue("Trace should explain the slope result", logContent.contains("reason=withinSlopeRange"));
    }

    @Test
    public void testSerializationRoundTrip() {
        RuleSerializationRoundTripTestSupport.assertRuleRoundTrips(series, rulePositiveSlope);
        RuleSerializationRoundTripTestSupport.assertRuleJsonRoundTrips(series, rulePositiveSlope);
        RuleSerializationRoundTripTestSupport.assertRuleRoundTrips(series, ruleNegativeSlope);
        RuleSerializationRoundTripTestSupport.assertRuleJsonRoundTrips(series, ruleNegativeSlope);
    }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/rules/InSlopeRuleTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.rules

import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.ta4j.core.BarSeries
import org.ta4j.core.BaseBarSeries
import org.ta4j.core.Indicator
import org.ta4j.core.indicators.helpers.FixedDecimalIndicator
import org.ta4j.core.num.Num

class InSlopeRuleTest {
    private var rulePositiveSlope: InSlopeRule? = null
    private var ruleNegativeSlope: InSlopeRule? = null
    @Before
    fun setUp() {
        val series: BarSeries = BaseBarSeries()
        val indicator: Indicator<Num> =
            FixedDecimalIndicator(series, 50.0, 70.0, 80.0, 90.0, 99.0, 60.0, 30.0, 20.0, 10.0, 0.0)
        rulePositiveSlope = InSlopeRule(indicator, series.numOf(20), series.numOf(30))
        ruleNegativeSlope = InSlopeRule(indicator, series.numOf(-40), series.numOf(-20))
    }

    @Test
    fun isSatisfied() {
            Assert.assertFalse(rulePositiveSlope!!.isSatisfied(0))
            Assert.assertTrue(rulePositiveSlope!!.isSatisfied(1))
            Assert.assertFalse(rulePositiveSlope!!.isSatisfied(2))
            Assert.assertFalse(rulePositiveSlope!!.isSatisfied(9))
            Assert.assertFalse(ruleNegativeSlope!!.isSatisfied(0))
            Assert.assertFalse(ruleNegativeSlope!!.isSatisfied(1))
            Assert.assertTrue(ruleNegativeSlope!!.isSatisfied(5))
            Assert.assertFalse(ruleNegativeSlope!!.isSatisfied(9))
        }
}